src/extra/components/xgm.py

```diff
@@ -391,19 +394,21 @@ def npulses(self, sase=None) -> int:
         """
         pg = self._check_sase_arg(sase)
         if pg not in self._npulses:
-            if pg == PropertyGroup.MAIN:
-                key = self._get_main_nbunches_key()
-            else:
-                key = f"pulseEnergy.numberOfSa{pg.value}BunchesActual"
-            self._npulses[pg] = int(self.control_source[key].as_single_value())
+            pulse_counts = self.pulse_counts(sase=sase)
+            if not np.allclose(pulse_counts[0], pulse_counts):
+                raise ValueError("Number of pulses is changing, there is no nominal number.")
+
+            self._npulses[pg] = int(pulse_counts[0])
 
         return self._npulses[pg]
 
     def pulse_counts(self, sase=None):
         """Return a 1D [DataArray][xarray.DataArray] of the number of pulses in each train.
 
-        See the docs for [XGM.npulses()][extra.components.XGM.npulses]
-        for information on retrieving the true number of pulses.
+        Warning:
+            This can be unreliable, see the docs for
+            [XGM.npulses()][extra.components.XGM.npulses] for information on
```

Comment:
For someone in a hurry to get something done, and maybe not a native English speaker, I think it would be easy to read this as 'this method is bad, `npulses()` is better'.

I think it might be worth repeating the pointer to `XrayPulses` in each of these warnings, even if that's a bit more verbose, rather than pointing people to the docs of another method to find that information.